### ISSUE TITLE
Breaking changes in FHIR 1.0.2 modelinfo

### ISSUE THREAD

**Issue by cmoesel** [cqframework/clinical_quality_language#240]
Previously, CQL could reference an Observation's value as a quantity via the field `valueQuantity` (e.g., `Observation O where O.valueQuantity < 40 'mg/dL'`).

Since that time, the `fhir-modelinfo-1.0.2.xml` definition has changed to support the new choice options, and the previous way of referencing a value quantity is no longer valid.  Now it seems that you just reference `value` and the CQL-to-ELM will cast appropriately if needed.

While this is easier for authoring (in some ways), it does reflect a departure from the way that FHIR is actually serialized on the wire -- you would never see a `value` field in FHIR data -- it would always be `valueQuantity` or `valueCodeableConcept`, etc.  From an implementation point of view, I found the more specific field names easier to deal with.  (Furthermore, I'm not sure you even see `value` in the FHIR spec, isn't it always `value[x]`?).  Is this representation up for discussion or is set in stone?

This also brings up an important point regarding versioning.  Our CDS artifacts on CDS Connect were compiled against the _old_ version of the `fhir-modelinfo-1.0.2.xml` file -- and thus use `valueQuantity`, `effectiveDateTime`, etc.  They will not compile against the _new_ version of that file.  The problem is, the modelinfo version number is tied to the FHIR version number (1.0.2), so even though significant _breaking_ changes were made to the modelinfo file, the version number was not incremented -- and as a result there is no way for the CDS Connect artifacts to indicate they used the _previous_ version of that modelinfo.  Worse yet, there are now two very different versions of the FHIR modelinfo version 1.0.2 floating around.

I think that the FHIR modelinfos (and perhaps others as well) need to somehow support versioning independent of the version of what they're modeling.  This means putting the FHIR version number in the model _name_ and starting at version _1.0_ of the modelinfo, or needing to introduce a 4th component to the version number, where components 1-3 match the FHIR version, but component 4 is incremented w/ changes to the modelinfo file (but at that point we are breaking semantic versioning).

**Comment by brynrhodes** [comment]
As part of the functionality introduced in CQL 1.2, we added the ability to represent choice types in CQL, and with the release of that version of the translator, models that could take advantage of that functionality began to. You're right that it's a breaking change though, in that we only offer the choice type, not the explicitly-typed attributes. 

I'm not excited about the prospect of adding a version number for the model info that is independent of the underlying model's version, we explicitly avoided that to avoid the added confusion that would inevitably introduce. However, I agree that we broke the versioning with the change we made and that needs to be addressed.

Option 1 to address that then is to roll back the changes (go back to an attribute that includes the type name), and then add a new model info, say FHIRWithChoices, that exposes the FHIR structures with the choice types instead of the type names.

Option 2 to address it is to add both to the model, so whenever there is an element that has value[x] representation in FHIR, expose both a "value" element of type Choice, and the typed-names, so that it supports both in the same model.

For QDM, we opted to only expose the Choice typed elements for simplicity, and so I replicated that behavior for the FHIR model infos.

Thoughts on which option would be better for supporting the work you're doing now, as well as moving forward?

**Comment by cmoesel** [comment]
That's a tough choice.

Option 1 would mean that we have two different, parallel modelinfos to represent the same underlying data model.  This means that one FHIR 1.0.2 artifact might look different than another FHIR 1.0.2 artifact -- and they wouldn't both necessarily be compatible with the same systems.  This could be confusing to users and implementers, and implementor code might be more complex to support both use cases.

Option 2 provides a consistent modelinfo, but introduces multiple ways to talk about the same data.  These choices also tend to confuse users and burden implementers (sigh).

So between those two, I don't know.

I wonder if there is a way to get the benefits of Option 2, but in a more consistent (and less duplicative) way?  The modelinfo already has a notion of `label` -- so perhaps we could use that to our advantage to support labels on choice options allowing them to be directly referenced in the CQL?

```xml
<ns4:element name="effective">
    <ns4:typeSpecifier xsi:type="ns4:ChoiceTypeSpecifier">
        <ns4:choice xsi:type="ns4:NamedTypeSpecifier" modelName="FHIR" name="dateTime" label="effectiveDateTime"/>
        <ns4:choice xsi:type="ns4:NamedTypeSpecifier" modelName="FHIR" name="Period" label="effectivePeriod"/>
    </ns4:typeSpecifier>
</ns4:element>
```

Then the CQL-to-ELM could still provide a consistent translation for implementors (eliminating any additional implementor burden).  What do you think of that?  Or does it introduce too much complexity to modelinfo itself?

Regarding the notion that we shouldn't independently version the modelinfo, I must disagree.  Without independently versioning it, there's no way to know what version someone built an artifact on without analyzing that artifact and how it uses the modelinfo.  That's fine if the *first* version of every modelinfo we write is 100% correct and we _never_ have to update it or fix a bug -- but is that a realistic expectation?  If we ever have to fix a bug in the modelinfo (or add a new feature), then this introduces a problem -- the CQL can't be deterministically compiled to ELM and the system implementor may receive class names or attributes that they don't recognize.  Avoiding these types of problems is the whole reason we have versioning in software and specifications, isn't it?  Isn't an instance of a modelinfo a type of specification itself that should be versioned like all other specifications?

**Comment by brynrhodes** [comment]
I agree, it's a touch choice, but your response leads me to offer an option 3:

Back out the changes for DSTU2 so that we leave it as is (i.e. still uses type-named attributes, not Choices), but use Choices for STU3. That addresses the issues for DSTU2 artifacts, but still allows the use of Choice elements in STU3 artifacts going forward.

I'm open to experimenting with using the notion of Labels, but for Choice types, the choices are defined using a TypeSpecifier, which does not have a Label, so this would represent a change to the ModelInfo, and an expansion of the type system, in that a "type specifier" would no longer be a way of just referencing a type, it would carry additional modifying information that was relevant during the process of path resolution.

As far as the modelinfo versioning, I disagree. The modelinfo doesn't own the versioning of the model, it's strictly a description of the model. If there are any errors in it (i.e. places in the model info that incorrectly describe the actual model), then artifacts built against that "version" of the model info either don't reference any of the incorrect descriptors, and therefore would build fine against a corrected model, or they reference incorrect descriptors and are therefore invalid artifacts that would fail at runtime. In other words, introducing an additional version for the model info would add complexity, but no value.

**Comment by cmoesel** [comment]
If we don't think others have adopted the DSTU2 model since it's been merged into master, then I'm fine with rolling it back -- and that would make our artifacts in CDS Connect compliant.  If we think others have adopted the new one though, I don't want to put them into the same situation I am -- so in that case we would want to offer two variants (but for DSTU2 only -- not STU3).

As for labels, it was just an idea for consideration.  You're right though -- the labeling would have to be applied to type specifiers, and that would introduce a new pattern.  I was trying to figure out how we make DSTU2 support both the new approach and old approach at the same time -- but if we rollback the DSTU2 model, then that's less important -- and we can go with standard choices for STU3, as you suggest.

As for modelinfo versioning, I guess we just have different opinions.  While I understand what you're saying, even in the situation you described (a bug that is fixed in the modelinfo), without versioning it's difficult to identify _which_ artifacts are potentially susceptible to it (e.g., "was this artifact built against the fixed or unfixed version?").  In addition, it means we should _never_ do what you tried to do with DSTU2; there is no opportunity for non-bug-fix enhancements or feature additions -- because you would run the risk of breaking artifacts (or implementations) that don't need to be broken.

**Comment by brynrhodes** [comment]
Included in 1.2.15
